File: auth/src/main/java/com/ykb/app/cryptotrader/auth/component/filter/HeaderFilter.java
package com.ykb.app.cryptotrader.auth.component.filter;

import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpFilter;
import jakarta.servlet.http.HttpServletResponse;
import jakarta.servlet.http.HttpServletRequest;
import org.springframework.security.authentication.AuthenticationServiceException;
import org.springframework.security.core.AuthenticationException;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.core.context.SecurityContextHolderStrategy;
import org.springframework.security.web.AuthenticationEntryPoint;
import org.springframework.security.web.savedrequest.RequestCache;

import java.io.IOException;

public abstract class HeaderFilter extends HttpFilter {

    private final AuthenticationEntryPoint authenticationEntryPoint;
    private final RequestCache requestCache;
    private final SecurityContextHolderStrategy securityContextHolderStrategy = SecurityContextHolder.getContextHolderStrategy();

    public HeaderFilter(AuthenticationEntryPoint authenticationEntryPoint, RequestCache requestCache) {
        this.authenticationEntryPoint=authenticationEntryPoint;
        this.requestCache=requestCache;
    }

    protected final void handleException(HttpServletRequest request, HttpServletResponse response, Exception reason) throws ServletException, IOException {
        var ex = reason;
        if(!(ex instanceof AuthenticationException)) {
            ex = new AuthenticationServiceException("Unexpected error occurred", ex);
        }

        var context = securityContextHolderStrategy.createEmptyContext();
        securityContextHolderStrategy.setContext(context);
        requestCache.saveRequest(request, response);
        authenticationEntryPoint.commence(request, response, (AuthenticationException) ex);
    }

}

File: library/auth/src/main/kotlin/com/ykb/app/cryptotrader/auth/filter/HeaderFilter.kt
package com.ykb.app.cryptotrader.auth.filter

import com.ykb.app.cryptotrader.utils.exceptions.auth.InternalServerErrorException
import jakarta.servlet.FilterChain
import jakarta.servlet.http.HttpFilter
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.security.core.AuthenticationException
import org.springframework.security.core.context.SecurityContext
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.security.core.context.SecurityContextHolderStrategy
import org.springframework.security.web.AuthenticationEntryPoint
import org.springframework.security.web.savedrequest.RequestCache

abstract class HeaderFilter : HttpFilter() {

    @Autowired
    private lateinit var authenticationEntryPoint: AuthenticationEntryPoint

    @Autowired
    private lateinit var requestCache: RequestCache

    private val securityContextHolderStrategy: SecurityContextHolderStrategy? = SecurityContextHolder.getContextHolderStrategy()

    protected fun handleException(request: HttpServletRequest?, response: HttpServletResponse?, chain: FilterChain?, reason: Exception) {
        var ex = reason
        ex.printStackTrace()
        if(ex !is AuthenticationException) {
            var message = ex.message
            if(message.isNullOrEmpty())
                message = "Unexpected error occurred"
            ex = InternalServerErrorException(message)
        }

        val context: SecurityContext = this.securityContextHolderStrategy!!.createEmptyContext()
        this.securityContextHolderStrategy.context = context
        this.requestCache.saveRequest(request, response)
        this.authenticationEntryPoint.commence(request, response, ex)
    }

}